### PR TITLE
Add support for first gemm scaling in attention kernels

### PR DESCRIPTION
**Pull request by manupak**
This commit adds support for scaling the first gemm output of attention kernels.
Additionally support for splat constant scale tensors are also added as they are quite common
with migraphx cases we are interested in supporting.

closes : https://github.com/ROCmSoftwarePlatform/rocMLIR-internal/issues/1094

### CODE REVIEW

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp [hunk 1]
```diff
@@ -1425,7 +1500,24 @@ struct GridwiseAttentionAccelRewritePattern
...
+                          "than gemm1LDSByteBufferSize.");
+    }
+    auto gemm1LDSByteBufferAType =
+        MemRefType::get({gemm1LDSByteBufferSize}, rewriter.getI8Type(),
+                        AffineMap{}, workgroupMemoryAddressSpace);
+    Value gemm1LDSByteBufferA = rewriter.create<memref::SubViewOp>(
+        loc, gemm1LDSByteBufferAType, ldsReductionWorkspaceByteBuffer,
+        ArrayRef<int64_t>{0}, ArrayRef<int64_t>{gemm1LDSByteBufferSize},
+        ArrayRef<int64_t>{1});
+    // Value gemm1LDSByteBufferA = ldsReductionWorkspaceByteBuffer;
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: Why commented out here?

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp [hunk 2]
```diff
@@ -1303,6 +1303,74 @@ struct GridwiseAttentionAccelRewritePattern
...
+  }
+
+  FailureOr<TypedAttr>
+  getSplatGlobalConstant(memref::GetGlobalOp getGlobalOp) const {
+    auto *symbolTableOp =
+        getGlobalOp->getParentWithTrait<OpTrait::SymbolTable>();
+    if (!symbolTableOp)
+      return failure();
+    auto global = dyn_cast_or_null<memref::GlobalOp>(
+        SymbolTable::lookupSymbolIn(symbolTableOp, getGlobalOp.getNameAttr()));
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: I think there's a way to look up the nearest symbol, you might want to use that?

**Review by krzysz00 — COMMENTED** [review]
As an overall comment, what happens if I pass in a scale that isn't a splat? Does it just get silently ignored?

**Comment by manupak** [comment]
> As an overall comment, what happens if I pass in a scale that isn't a splat? Does it just get silently ignored?

So currently we have an open issue about handling bound non-scalar constants : https://github.com/ROCmSoftwarePlatform/rocMLIR-internal/issues/1171

It is not still clear to me do we want to support non-scalar constants bound to kernels or leave them to host to be supplied.
MIGraphX will not bind these as of now.

No, it does not get silently ignored; it will just fail the pass as unsupported.
But non-scalar args will be handled as long as kernel is concerned.

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp (hunk 2 above)
Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: any pointers would be appreciated.

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp (hunk 1 above)
Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: forgot to delete

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp (hunk 2 above)
Comment (krzysz00) [reply mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: `lookupNearestSymbolFrom` is a static method in `SymbolTable`

Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: thanks! now changed

**Review by krzysz00 — APPROVED** [review]

**Pull request merged**